File: scalechain-storage/src/main/scala/io/scalechain/blockchain/storage/index/TransactionDescriptorIndex.scala
package io.scalechain.blockchain.storage.index

import io.scalechain.blockchain.proto.{TransactionDescriptor, Hash}
import io.scalechain.blockchain.proto.codec.{TransactionDescriptorCodec, TransactionPoolEntryCodec, HashCodec}

trait TransactionDescriptorIndex {
  import DatabaseTablePrefixes._
  private implicit val hashCodec = HashCodec
  private implicit val transactionCodec = TransactionPoolEntryCodec

  /**
    * Get the descriptor of a transaction by hash
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash.
    * @return Some(descriptor) if the transaction exists; None otherwise.
    */
  def getTransactionDescriptor(txHash : Hash)(implicit db : KeyValueDatabase) : Option[TransactionDescriptor] = {
    db.getObject(TRANSACTION, txHash)(HashCodec, TransactionDescriptorCodec)
  }

  /**
    * Put the descriptor of a transaction with hash of it
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash.
    * @param transactionDescriptor The descriptor of the transaction.
    */
  def putTransactionDescriptor(txHash : Hash, transactionDescriptor : TransactionDescriptor)(implicit db : KeyValueDatabase) = {
    db.putObject(TRANSACTION, txHash, transactionDescriptor)(HashCodec, TransactionDescriptorCodec)
  }

  /**
    * Del the descriptor of a transaction by hash.
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash
    */
  def delTransactionDescriptor(txHash : Hash)(implicit db : KeyValueDatabase) : Unit = {
    db.delObject(TRANSACTION, txHash)(HashCodec)
  }
}

File: scalechain-storage/src/main/kotlin/io/scalechain/blockchain/storage/index/TransactionDescriptorIndex.kt
package io.scalechain.blockchain.storage.index

import io.scalechain.blockchain.proto.TransactionDescriptor
import io.scalechain.blockchain.proto.Hash
import io.scalechain.blockchain.proto.codec.TransactionDescriptorCodec
import io.scalechain.blockchain.proto.codec.HashCodec

interface TransactionDescriptorIndex {
//  private val logger = LoggerFactory.getLogger(TransactionDescriptorIndex::class.java)

  /**
    * Get the descriptor of a transaction by hash
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash.
    * @return Some(descriptor) if the transaction exists; None otherwise.
    */
  fun getTransactionDescriptor(db : KeyValueDatabase, txHash : Hash) : TransactionDescriptor? {
    //logger.trace(s"getTransactionDescriptor : ${txHash}")
    return db.getObject(HashCodec, TransactionDescriptorCodec, DB.TRANSACTION, txHash)
  }

  /**
    * Put the descriptor of a transaction with hash of it
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash.
    * @param transactionDescriptor The descriptor of the transaction.
    */
  fun putTransactionDescriptor(db : KeyValueDatabase, txHash : Hash, transactionDescriptor : TransactionDescriptor) : Unit {
    //logger.trace(s"putTransactionDescriptor : ${txHash}")
    db.putObject(HashCodec, TransactionDescriptorCodec, DB.TRANSACTION, txHash, transactionDescriptor)
  }

  /**
    * Del the descriptor of a transaction by hash.
    *
    * TODO : Add a test case
    *
    * @param txHash The transaction hash
    */
  fun delTransactionDescriptor(db : KeyValueDatabase, txHash : Hash) : Unit {
    //logger.trace(s"delTransactionDescriptor : ${txHash}")
    db.delObject(HashCodec, DB.TRANSACTION, txHash)
  }
}